consensus/ibft/sign.go

```diff
@@ -184,35 +172,75 @@ func verifyCommittedFields(
 		return fmt.Errorf("empty committed seals")
 	}
 
+	// get the message that needs to be signed
+	// this not signing! just removing the fields that should be signed
 	hash, err := calculateHeaderHash(header)
 	if err != nil {
 		return err
 	}
 
 	rawMsg := wrapCommitHash(hash)
 
+	return verifyCommittedSealsImpl(extra.CommittedSeal, rawMsg, snap.Set, quorumSizeFn)
+}
+
+// verifyCommittedFields verifies ParentCommittedSeal in IBFT Extra of Header
+func verifyParentCommittedSeal(
+	parentSnap *Snapshot,
+	parent, header *types.Header,
+	quorumSizeFn QuorumImplementation,
+) error {
+	parentCommittedSeals, err := unpackParentCommittedSealFromIbftExtra(header)
+	if err != nil {
+		return err
+	}
+
+	hash, err := calculateHeaderHash(parent)
+	if err != nil {
+		return err
+	}
+
+	rawMsg := wrapCommitHash(hash)
+
+	return verifyCommittedSealsImpl(parentCommittedSeals, rawMsg, parentSnap.Set, quorumSizeFn)
+}
+
+// verifyCommittedSealsImpl verifies given committedSeals
+// checks committedSeals are the correct signatures signed by the validators and the number of seals
+func verifyCommittedSealsImpl(
+	committedSeals [][]byte,
+	msg []byte,
+	validators ValidatorSet,
+	quorumSizeFn QuorumImplementation,
+) error {
+	// Committed seals shouldn't be empty
+	if len(committedSeals) == 0 {
+		return ErrEmptyCommittedSeals
+	}
+
 	visited := map[types.Address]struct{}{}
 
-	for _, seal := range extra.CommittedSeal {
-		addr, err := ecrecoverImpl(seal, rawMsg)
+	for _, seal := range committedSeals {
+		addr, err := ecrecoverImpl(seal, msg)
 		if err != nil {
 			return err
 		}
 
 		if _, ok := visited[addr]; ok {
-			return fmt.Errorf("repeated seal")
-		} else {
-			if !snap.Set.Includes(addr) {
-				return fmt.Errorf("signed by non validator")
-			}
-			visited[addr] = struct{}{}
+			return ErrRepeatedCommittedSeal
+		}
+
+		if !validators.Includes(addr) {
+			return ErrNonValidatorCommittedSeal
 		}
+
+		visited[addr] = struct{}{}
 	}
 
 	// Valid committed seals must be at least 2F+1
 	// 2F 	is the required number of honest validators who provided the committed seals
 	// +1	is the proposer
-	if validSeals := len(visited); validSeals < quorumSizeFn(snap.Set) {
+	if validSeals := len(visited); validSeals < quorumSizeFn(validators) {
 		return fmt.Errorf("not enough seals to seal block")
```

Comment:
we have a const for this err: ErrNotEnoughCommittedSeals

Reply:
Resolved in the latest PR